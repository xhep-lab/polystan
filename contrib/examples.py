"""
Find example programs
=====================
"""

import os


CWD = os.path.dirname(os.path.realpath(__file__))
ROOT = os.path.normpath(os.path.join(CWD, ".."))
EXAMPLE_DIR = os.path.join(ROOT, "examples")


def search_programs(dir_):
    return [
        os.path.join(dir_, n)
        for n in os.listdir(dir_)
        if n.endswith(".stan")
    ]


def examples():
    return search_programs(EXAMPLE_DIR)


def example(name):
    stan_file = os.path.join(EXAMPLE_DIR, name)
    assert os.path.isfile(stan_file)
    return stan_file
