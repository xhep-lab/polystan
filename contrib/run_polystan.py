"""
Python wrapper to build & run polystan program
==============================================
"""

import os
import subprocess

import arviz as az

CWD = os.path.dirname(os.path.realpath(__file__))
ROOT = os.path.normpath(os.path.join(CWD, ".."))


def make_polystan(target):
    subprocess.check_call(f"make {target}", shell=True, cwd=ROOT)


def cli_subargs(**kwargs):
    return " ".join(f"--{k}={v}" for k, v in kwargs.items())


def cli_args(**kwargs):
    return " ".join(f"{k} {cli_subargs(**v)}" for k, v in kwargs.items())


def find_data_file(target):
    data_file = f"{target}.data.json"
    if os.path.isfile(data_file):
        return data_file
    return None


def run_polystan(stan_file, data_file=None, seed=0, **kwargs):

    target = os.path.splitext(stan_file)[0]

    make_polystan(target)

    args = {
        "random": {"seed": seed},
        "polychord": {
            "seed": seed,
            "overwrite": True,
        },
    }

    for k, v in kwargs.items():
        args[k].update(v)

    if data_file is None:
        data_file = find_data_file(target)

    if data_file is not None:
        args["data"] = {"file": data_file}

    subprocess.check_call(f"{target} {cli_args(**args)}", shell=True)

    name = os.path.split(target)[1]
    result_name = f"{name}.json"
    return az.from_json(result_name)
