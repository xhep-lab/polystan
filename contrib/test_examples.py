"""
Test example programs
=====================
"""

import pytest

from run_polystan import run_polystan
from examples import examples


TEST_SETTINGS = {"no-feedback": True, "no-write": True,
                 "no-derived": True, "nlive": 1000, "write-stats": True}


def read_evidence(data):
    return data['sample_stats']['evidence'].data[0][0]['log evidence']


@pytest.mark.parametrize("example", examples())
def test_evidence(example, snapshot):
    data = run_polystan(example, polychord=TEST_SETTINGS)
    assert read_evidence(data) == snapshot
