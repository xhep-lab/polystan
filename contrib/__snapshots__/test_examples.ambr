# serializer version: 1
# name: test_evidence[/home/andrew/repos/polystan/examples/bernoulli.stan]
  -6.199459075927734
# ---
# name: test_evidence[/home/andrew/repos/polystan/examples/disaster.stan]
  -176.7947540283203
# ---
# name: test_evidence[/home/andrew/repos/polystan/examples/eggbox.stan]
  -15.066864967346191
# ---
# name: test_evidence[/home/andrew/repos/polystan/examples/gaussian.stan]
  -0.00551773514598608
# ---
# name: test_evidence[/home/andrew/repos/polystan/examples/glmm_h0.stan]
  -156.30838012695312
# ---
# name: test_evidence[/home/andrew/repos/polystan/examples/glmm_h1.stan]
  -156.48731994628906
# ---
# name: test_evidence[/home/andrew/repos/polystan/examples/himmelblau.stan]
  -4.4977874755859375
# ---
# name: test_evidence[/home/andrew/repos/polystan/examples/priors.stan]
  -0.0009999993490055203
# ---
# name: test_evidence[/home/andrew/repos/polystan/examples/rastrigin.stan]
  -22.82518768310547
# ---
# name: test_evidence[/home/andrew/repos/polystan/examples/rosenbrock.stan]
  -9.264966011047363
# ---
# name: test_evidence[/home/andrew/repos/polystan/examples/shell.stan]
  -5.59080171585083
# ---
# name: test_evidence[/home/andrew/repos/polystan/examples/slab_spike.stan]
  -4.565235137939453
# ---
